crates/spirv-std/src/arch/ray_tracing.rs

```diff
@@ -0,0 +1,81 @@
+/// Reports an intersection back to the traversal infrastructure.
+///
+/// If the intersection occurred within the current ray interval, the
+/// intersection confirmation is performed (see the API specification for more
+/// details). If the value of Hit falls outside the current ray interval, the
+/// hit is rejected.
+///
+/// Returns True if the hit was accepted by the ray interval and the intersection was confirmed. Returns False otherwise.
+///
+/// - `hit` is the floating point parametric value along ray for the intersection.
+/// - `hit_kind` is the integer hit kind reported back to other shaders and
+///   accessible by the `hit kind` builtin.
+///
+/// This instruction is allowed only in IntersectionKHR execution model.
+///
+/// This instruction is a shader call instruction which may invoke shaders with
+/// the `any_hit` execution model.
+#[spirv_std_macros::gpu_only]
+#[doc(alias = "OpReportIntersectionKHR")]
+#[inline]
+pub unsafe fn report_intersection(hit: f32, hit_kind: u32) -> bool {
+    let mut result: u8 = 0;
+
+    asm! {
+        "%bool = OpTypeBool",
+        "%u8 = OpTypeInt 8 0",
+        "%result = OpReportIntersectionKHR %u8 {hit} {hit_kind}",
```

Comment:
Result type should be a `%bool` - from the spec:

> Result Type must be a scalar boolean.

Reply:
Yeah it should, but it doesn't seem to work. Using bool seems to insert an `OpINotEqual` instruction, which causes it error.

```rust
pub unsafe fn report_intersection(hit: f32, hit_kind: u32) -> bool {
    let result: u8;

    asm! {
        "%bool = OpTypeBool",
        "%u8 = OpTypeInt 8 0",
        "{result} = OpReportIntersectionKHR %bool {hit} {hit_kind}",
        result = out(reg) result,
        hit = in(reg) hit,
        hit_kind = in(reg) hit_kind,
    };

    result != 0
}
```

```spirv
; SPIR-V
; Version: 1.3
; Generator: Google rspirv; 0
; Bound: 17
; Schema: 0
               OpCapability Shader
               OpCapability Int8
               OpCapability RayTracingKHR
               OpExtension "SPV_KHR_ray_tracing"
               OpMemoryModel Logical Simple
               OpEntryPoint IntersectionNV %1 "main"
               OpName %report_intersection_khr__main "report_intersection_khr::main"
       %bool = OpTypeBool
      %float = OpTypeFloat 32
       %uint = OpTypeInt 32 0
      %uchar = OpTypeInt 8 0
    %uchar_0 = OpConstant %uchar 0
       %void = OpTypeVoid
          %9 = OpTypeFunction %void
    %float_2 = OpConstant %float 2
     %uint_4 = OpConstant %uint 4
%report_intersection_khr__main = OpFunction %void None %9
         %12 = OpLabel
         %13 = OpReportIntersectionKHR %bool %float_2 %uint_4
         %14 = OpINotEqual %bool %13 %uchar_0
               OpReturn
               OpFunctionEnd
          %1 = OpFunction %void None %9
         %15 = OpLabel
         %16 = OpFunctionCall %void %report_intersection_khr__main
               OpReturn
               OpFunctionEnd
```

Reply:
I'm confused, the spec isn't something you can ignore and make it be a u8 instead, the spec states

> Result Type **must** be a scalar boolean.

maybe I shouldn't have said "should", it _must_ be a boolean.

Reply:
This seems to work just fine

```rust
pub unsafe fn report_intersection(hit: f32, hit_kind: u32) -> bool {
    let result: u32;

    asm! {
        "%bool = OpTypeBool",
        "%u32 = OpTypeInt 32 0",
        "%zero = OpConstant %u32 0",
        "%one = OpConstant %u32 0",
        "%result = OpReportIntersectionKHR %bool {hit} {hit_kind}",
        "{result} = OpSelect %u32 %result %one %zero",
        result = out(reg) result,
        hit = in(reg) hit,
        hit_kind = in(reg) hit_kind,
    };

    result != 0
}
```